src/java/org/apache/cassandra/index/sai/plan/StorageAttachedIndexSearcher.java

```diff
@@ -379,6 +374,7 @@ private UnfilteredRowIterator queryStorageAndFilter(PrimaryKey key)
             try (UnfilteredRowIterator partition = queryController.queryStorage(key, executionController))
             {
                 queryContext.partitionsRead++;
+                queryContext.checkpoint();
```

Comment:
At the end of the day, we probably don't need any more checkpointing than this. Materializing and post-filtering a single row is a perfectly reasonable boundary unit for timeout checking.